[web-animations] imported/blink/fast/css-generated-content/pseudo-animation-display.html is a failure

graouts opened the pull request:
#### b8a264c14a4a53ba960db617db37dd97496811fc
<pre>
[web-animations] imported/blink/fast/css-generated-content/pseudo-animation-display.html is a failure
<a href="https://bugs.webkit.org/show_bug.cgi?id=271375">https://bugs.webkit.org/show_bug.cgi?id=271375</a>
<a href="https://rdar.apple.com/125156782">rdar://125156782</a>

Reviewed by Antti Koivisto.

Pseudo-elements represented by a `PseudoElement`, namely `::before` and `::after`, are torn down if
a `display: none` style is set. An exception to that rule is if such a pseudo-element is targeted
by a script-originated `Animation` object.

Now that the `display` property can be animated, it is possible that a `PseudoElement` is targeted
by an animation that will set `display: none`. So we now ensure that we no longer tear down such
`PseudoElement` objects.

To achieve this, first we move the `animationsAffectedDisplay` bit from `ElementUpdate` to `RenderStyle`,
renaming it as `hasDisplayAffectedByAnimations`, such that it is accessible throughout the style update
since pseudo-element style updates don&apos;t have access to the `ElementUpdate` yielded for that specific
pseudo-element, on the one yielded for their owner element.

Second, we also add a new static function `elementHasDisplayAnimationForPseudoId()` we we use within
`RenderTreeUpdater::GeneratedContent::updatePseudoElement()` to introduce another case where we don&apos;t
tear down `PseudoElement` objects if `display: none` is set: the case where such a pseudo-element is
targeted by an animation affecting the `display` property.

Finally, we can remove the now-outdated `imported/blink/fast/css-generated-content/pseudo-animation-display.html`
test and replace it with a new WPT test. That test is marked as tentative because I filed a CSS WG issue
(<a href="https://github.com/w3c/csswg-drafts/issues/10111)">https://github.com/w3c/csswg-drafts/issues/10111)</a> about whether the approach taken here is correct.

* LayoutTests/imported/blink/fast/css-generated-content/pseudo-animation-display-expected.txt: Removed.
* LayoutTests/imported/blink/fast/css-generated-content/pseudo-animation-display.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-dont-cancel-pseudo.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-dont-cancel-pseudo.tentative.html: Added.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasDisplayAffectedByAnimations const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setHasDisplayAffectedByAnimations):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
(WebCore::StyleMiscNonInheritedData::StyleMiscNonInheritedData):
(WebCore::StyleMiscNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::keyframeEffectStackForElementAndPseudoId):
(WebCore::elementIsTargetedByKeyframeEffectRequiringPseudoElement):
(WebCore::elementHasDisplayAnimationForPseudoId):
(WebCore::RenderTreeUpdater::GeneratedContent::updatePseudoElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/StyleUpdate.h:

Canonical link: <a href="https://commits.webkit.org/276568@main">https://commits.webkit.org/276568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19fbc5614cb75e86df399ddfba47b1c6e91e8d3c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44973 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24082 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47472 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47631 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40980 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/47279 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21480 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36919 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45551 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21173 "Failed to checkout and rebase branch from PR 26312") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38731 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18006 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39862 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3021 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/41276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40164 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49303 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19945 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16492 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43914 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21261 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38955 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42683 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6260 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20940 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
```diff
@@ -90,6 +90,7 @@ class StyleMiscNonInheritedData : public RefCounted<StyleMiscNonInheritedData> {
     LengthPoint objectPosition;
     int order;
 
+    unsigned animationsAffectedDisplay : 1 { false };
```
hasDisplayAffectedByAnimations? Same for accessors.

graouts replied:
Will fix in commit.

anttijk commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -421,7 +421,7 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
         // display:none cancels animations.
         auto teardownType = [&]() {
-            if (!elementUpdate.animationsAffectedDisplay && elementUpdate.style->display() == DisplayType::None)
+            if (!elementUpdate.style->animationsAffectedDisplay() && elementUpdate.style->display() == DisplayType::None)
```
Only place I see reading animationsAffectedDisplay bit is here? Why is it needed for pseudo elements?

graouts committed:
[web-animations] imported/blink/fast/css-generated-content/pseudo-animation-display.html is a failure
https://bugs.webkit.org/show_bug.cgi?id=271375
rdar://125156782

Reviewed by Antti Koivisto.

Pseudo-elements represented by a `PseudoElement`, namely `::before` and `::after`, are torn down if
a `display: none` style is set. An exception to that rule is if such a pseudo-element is targeted
by a script-originated `Animation` object.

Now that the `display` property can be animated, it is possible that a `PseudoElement` is targeted
by an animation that will set `display: none`. So we now ensure that we no longer tear down such
`PseudoElement` objects.

To achieve this, first we move the `animationsAffectedDisplay` bit from `ElementUpdate` to `RenderStyle`,
renaming it as `hasDisplayAffectedByAnimations`, such that it is accessible throughout the style update
since pseudo-element style updates don't have access to the `ElementUpdate` yielded for that specific
pseudo-element, on the one yielded for their owner element.

Second, we also add a new static function `elementHasDisplayAnimationForPseudoId()` we we use within
`RenderTreeUpdater::GeneratedContent::updatePseudoElement()` to introduce another case where we don't
tear down `PseudoElement` objects if `display: none` is set: the case where such a pseudo-element is
targeted by an animation affecting the `display` property.

Finally, we can remove the now-outdated `imported/blink/fast/css-generated-content/pseudo-animation-display.html`
test and replace it with a new WPT test. That test is marked as tentative because I filed a CSS WG issue
(https://github.com/w3c/csswg-drafts/issues/10111) about whether the approach taken here is correct.

* LayoutTests/imported/blink/fast/css-generated-content/pseudo-animation-display-expected.txt: Removed.
* LayoutTests/imported/blink/fast/css-generated-content/pseudo-animation-display.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-dont-cancel-pseudo.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-animations/display-none-dont-cancel-pseudo.tentative.html: Added.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::hasDisplayAffectedByAnimations const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setHasDisplayAffectedByAnimations):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
(WebCore::StyleMiscNonInheritedData::StyleMiscNonInheritedData):
(WebCore::StyleMiscNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):
* Source/WebCore/rendering/updating/RenderTreeUpdaterGeneratedContent.cpp:
(WebCore::keyframeEffectStackForElementAndPseudoId):
(WebCore::elementIsTargetedByKeyframeEffectRequiringPseudoElement):
(WebCore::elementHasDisplayAnimationForPseudoId):
(WebCore::RenderTreeUpdater::GeneratedContent::updatePseudoElement):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
* Source/WebCore/style/StyleUpdate.h:

Canonical link: https://commits.webkit.org/276568@main
